company-lsp.el

```diff
@@ -305,22 +328,15 @@ See the documentation of `company-backends' for COMMAND and ARG."
              (lsp--capability "completionProvider")
              (or (company-lsp--completion-prefix) 'stop)))
     (candidates
-     (cons :async
-           #'(lambda (callback)
-               (let ((cache (company-lsp--cache-get arg)))
-                 (if cache
-                     (funcall callback (plist-get cache :candidates))
-                   (let ((req (lsp--make-request "textDocument/completion"
-                                                 (lsp--text-document-position-params))))
-                     (setq company-lsp--line-backup
-                           (buffer-substring (line-beginning-position) (line-end-position)))
-                     (if company-lsp-async
-                         (lsp--send-request-async req
-                                                  (lambda (resp)
-                                                    (company-lsp--on-completion resp arg callback)))
-                       (company-lsp--on-completion (lsp--send-request req)
-                                                   arg
-                                                   callback))))))))
+     (setq company-lsp--line-backup
+           (buffer-substring (line-beginning-position) (line-end-position)))
+     (let ((cache (company-lsp--cache-get arg)))
+       (if cache
+           (plist-get cache :candidates)
+         (if company-lsp-async
+             (cons :async (lambda (callback)
+                            (company-lsp--candidates-async arg callback)))
+           (company-lsp--candidates-sync arg)))))
```

Comment:
Thanks for asking. Looks good.
I would just write:
```el
(or (plist-get (company-lsp--cache-get arg) :candidates)
    (and company-lsp-async
         (cons :async (lambda (callback)
                        (company-lsp--candidates-async arg callback))))
    (company-lsp--candidates-sync arg))

```

Reply:
Thanks. Done.